crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -177,15 +179,9 @@ impl ClientState {
         })
     }
 
-    pub fn with_header(self, h: TmHeader) -> Result<Self, Error> {
+    pub fn with_header(self, header: TmHeader) -> Result<Self, Error> {
         Ok(ClientState {
-            latest_height: Height::new(
-                self.latest_height.revision_number(),
-                h.signed_header.header.height.into(),
-            )
-            .map_err(|_| Error::InvalidHeaderHeight {
-                height: h.signed_header.header.height.value(),
-            })?,
+            latest_height: max(header.height(), self.latest_height),
```

Comment:
> Creating a new `ClientState` without using `new` method skips the `latest_height` checks.

It's not possible to create a `ClientState` object outside the module without calling `new()` because `ClientState` has some private fields. So we know the checks will always run

Reply:
> (On a side note, this is one of the reasons I generally advocate separating new instance creation and validation methods and prefer IBC-go [approach](https://github.com/cosmos/ibc-go/blob/5a67efc496125823f6ca998048ce6c95dcc98427/modules/light-clients/07-tendermint/client_state.go#L112) in this regard)

I consider that a code smell in most cases. Invariants about an object field should be checked as you're constructing an object; otherwise in your code you never know if an object you receive is valid. It's just another source of bugs. A similar concept in C++ is [RAII](https://en.cppreference.com/w/cpp/language/raii) (basically objects constructed should always be valid and ready to use without needing to call another function after the constructor).

One scenario where we might want to do checks outside the constructor though is if some invariant checks are very expensive, and we only want to run them in `validate()` (and not `execute()`). Otherwise I think we should stick to the general principle above.

Reply:
> It's not possible to create a ClientState object outside the module without calling new() because ClientState has some private fields. So we know the checks will always run

Right. By the way, good to have a note here so that calling `with_header` (in case needed later anywhere else in the repo, specifically for tests) should be done with caution.

Reply:
> I consider that a code smell in most cases. Invariants about an object field should be checked as you're constructing an object; otherwise in your code you never know if an object you receive is valid. It's just another source of bugs. A similar concept in C++ is [RAII](https://en.cppreference.com/w/cpp/language/raii) (basically objects constructed should always be valid and ready to use without needing to call another function after the constructor).

> One scenario where we might want to do checks outside the constructor though is if some invariant checks are very expensive, and we only want to run them in validate() (and not execute()). Otherwise I think we should stick to the general principle above.

I see, that's another way to look at it. Though, already IBC-go is doing that differently. 
Nevertheless, for now, we don't really need to worry about it. We came up with a good context here to keep in mind, in case needed to take a closer look. 🙏🏻

Reply:
> Though, already IBC-go is doing that differently.

Note that we mainly use ibc-go to infer what the IBC protocol is (since the standard is out of date/inaccurate, as previously stated). The patterns used in the code and general code style are irrelevant to us. We sometimes borrow some architectural ideas (such as what we did with this PR), but that's about it.